table/block_based/block_based_table_builder.cc

```diff
@@ -1501,24 +1501,28 @@ Status BlockBasedTableBuilder::InsertBlockInCache(const Slice& block_contents,
     const size_t read_amp_bytes_per_bit =
         rep_->table_options.read_amp_bytes_per_bit;
 
-    TBlocklike* block_holder = BlocklikeTraits<TBlocklike>::Create(
-        std::move(results), read_amp_bytes_per_bit,
-        rep_->ioptions.statistics.get(),
-        false /*rep_->blocks_definitely_zstd_compressed*/,
-        rep_->table_options.filter_policy.get());
-
-    if (block_holder->own_bytes()) {
-      size_t charge = block_holder->ApproximateMemoryUsage();
-      s = block_cache->Insert(key, block_holder, charge,
-                              &DeleteEntryCached<TBlocklike>);
-
-      if (s.ok()) {
-        BlockBasedTable::UpdateCacheInsertionMetrics(
-            block_type, nullptr /*get_context*/, charge, s.IsOkOverwritten(),
-            rep_->ioptions.stats);
-      } else {
-        RecordTick(rep_->ioptions.stats, BLOCK_CACHE_ADD_FAILURES);
-      }
+    // TODO akanksha:: Dedup below code by calling
+    // BlockBasedTable::PutDataBlockToCache.
+    std::unique_ptr<TBlocklike> block_holder(
+        BlocklikeTraits<TBlocklike>::Create(
+            std::move(results), read_amp_bytes_per_bit,
+            rep_->ioptions.statistics.get(),
+            false /*rep_->blocks_definitely_zstd_compressed*/,
+            rep_->table_options.filter_policy.get()));
+
+    assert(block_holder->own_bytes());
+    size_t charge = block_holder->ApproximateMemoryUsage();
+    s = block_cache->Insert(
+        key, block_holder.get(),
```

Comment:
I don't think it's fixed yet. (Now use-after-free or double-delete on Insert OK, but Cache::DisownData() means it won't be caught unless running under ASAN or Valgrind.)